pyccel/parser/semantic.py

```diff
@@ -988,7 +990,9 @@ def _assign_lhs_variable(self, lhs, d_var, rhs, new_expressions, is_augassign, *
 
                 # ...
                 # Add memory allocation if needed
-                if lhs.on_heap:
+                array_declared_in_function = isinstance(rhs, FunctionCall) and not isinstance(rhs.funcdef, PyccelFunctionDef) \
+                                            and not rhs.funcdef.is_elemental and not isinstance(lhs, HomogeneousTupleVariable)
```

Comment:
This needs expanding to handle the case where the array is declared as the result of a function returning multiple elements (called from here : https://github.com/pyccel/pyccel/blob/master/pyccel/parser/semantic.py#L2423 or here : https://github.com/pyccel/pyccel/blob/master/pyccel/parser/semantic.py#L2431-L2432 )

Reply:
I'm not sure if it can be done with the information currently passed to the function. You may need to add an extra argument